x-pack/plugins/secops/public/components/timeline/index.tsx

```diff
@@ -4,82 +4,143 @@
-  display: flex;
-  flex-direction: column;
-  min-height: 700px;
-  overflow: none;
-  user-select: none;
-  width: ${props => props.width};
-  height: ${props => props.height};
-`;
-
-const defaultHeight = '100%';
-
-/** The parent Timeline component */
-export const Timeline = pure<Props>(
-  ({
-    columnHeaders,
-    columnRenderers,
-    dataProviders,
-    data,
-    height = defaultHeight,
-    onColumnSorted,
-    onDataProviderRemoved,
-    onFilterChange,
-    onRangeSelected,
-    rowRenderers,
-    sort,
-    width,
-  }) => (
-    <TimelineDiv data-test-subj="timeline" width={`${width}px`} height={height}>
-      <TimelineHeader
-        dataProviders={dataProviders}
-        onDataProviderRemoved={onDataProviderRemoved}
-        width={width}
-      />
-      <Body
-        columnHeaders={columnHeaders}
+interface DispatchProps {
+  createTimeline: ActionCreator<{ id: string }>;
+  addProvider: ActionCreator<{
+    id: string;
+    provider: DataProvider;
+  }>;
+  updateData: ActionCreator<{
+    id: string;
+    data: ECS[];
+  }>;
+  updateProviders: ActionCreator<{
+    id: string;
+    providers: DataProvider[];
+  }>;
+  updateRange: ActionCreator<{
+    id: string;
+    range: Range;
+  }>;
+  updateSort: ActionCreator<{
+    id: string;
+    sort: Sort;
+  }>;
+  removeProvider: ActionCreator<{
+    id: string;
+    providerId: string;
+  }>;
+}
+
+type Props = OwnProps & StateProps & DispatchProps;
+
+class StatefulTimelineComponent extends React.PureComponent<Props> {
+  public componentDidMount() {
+    const { createTimeline, id } = this.props;
+
+    createTimeline({ id });
+  }
+
+  public render() {
+    const {
+      data,
+      dataProviders,
+      headers,
+      id,
+      range,
+      removeProvider,
+      sort,
+      updateRange,
+      updateSort,
+      width,
+    } = this.props;
+
+    const onColumnSorted: OnColumnSorted = sorted => {
+      updateSort({ id, sort: sorted });
+    };
+
+    const onDataProviderRemoved: OnDataProviderRemoved = dataProvider => {
+      removeProvider({ id, providerId: dataProvider.id });
+    };
+
+    const onRangeSelected: OnRangeSelected = selectedRange => {
+      updateRange({ id, range: selectedRange });
+    };
+
+    return (
+      <Timeline
+        columnHeaders={headers}
         columnRenderers={columnRenderers}
         dataProviders={dataProviders}
         data={data}
         onColumnSorted={onColumnSorted}
         onDataProviderRemoved={onDataProviderRemoved}
-        onFilterChange={onFilterChange}
+        onFilterChange={noop}
         onRangeSelected={onRangeSelected}
+        range={range}
         rowRenderers={rowRenderers}
         sort={sort}
         width={width}
       />
-    </TimelineDiv>
-  )
-);
+    );
+  }
+}
+
+const mapStateToProps = (state: State, { id }: OwnProps) => {
+  const timeline = timelineByIdSelector(state)[id];
+  const { dataProviders, data, range, sort } = timeline || timelineDefaults;
+
+  return timeline != null
```

Comment:
```suggestion
  _.defaultTo({ id, dataProviders, data, sort }, timeline)
```

That would be clearer IMHO